pkg/capacityscheduling/capacity_scheduling.go

```diff
@@ -0,0 +1,816 @@
+			candidates = append(candidates, &c)
+			resultLock.Unlock()
+		}
+	}
+	pluginsutil.Until(ctx, len(potentialNodes), checkNode)
+	return candidates
+}
+
+// nodesWherePreemptionMightHelp returns a list of nodes with failed predicates
+// that may be satisfied by removing pods from the node.
+func nodesWherePreemptionMightHelp(nodes []*framework.NodeInfo, m framework.NodeToStatusMap) []*framework.NodeInfo {
+	var potentialNodes []*framework.NodeInfo
+	for _, node := range nodes {
+		name := node.Node().Name
+		// We reply on the status by each plugin - 'Unschedulable' or 'UnschedulableAndUnresolvable'
+		// to determine whether preemption may help or not on the node.
+		if m[name].Code() == framework.UnschedulableAndUnresolvable {
+			continue
+		}
+		potentialNodes = append(potentialNodes, node)
+	}
+	return potentialNodes
+}
+
+func selectVictimsOnNode(
+	ctx context.Context,
+	ph framework.PreemptHandle,
+	state *framework.CycleState,
+	pod *v1.Pod,
+	nodeInfo *framework.NodeInfo,
+	pdbs []*policy.PodDisruptionBudget,
+) ([]*v1.Pod, int, bool) {
+	elasticQuotaSnapshotState, err := getElasticQuotaSnapshotState(state)
+	if err != nil {
+		klog.Errorf("error reading %q from cycleState: %v", ElasticQuotaSnapshotKey, err)
+		return nil, 0, false
+	}
+
+	preFilterState, err := getPreFilterState(state)
+	if err != nil {
+		klog.Errorf("error reading %q from cycleState: %v", preFilterStateKey, err)
+		return nil, 0, false
+	}
+
+	var potentialVictims []*v1.Pod
+	var moreThanMinWithPreemptor bool
+
+	removePod := func(rp *v1.Pod) error {
+		if err := nodeInfo.RemovePod(rp); err != nil {
+			return err
+		}
+		status := ph.RunPreFilterExtensionRemovePod(ctx, state, pod, rp, nodeInfo)
+		if !status.IsSuccess() {
+			return status.AsError()
+		}
+		return nil
+	}
+
+	addPod := func(ap *v1.Pod) error {
+		nodeInfo.AddPod(ap)
+		status := ph.RunPreFilterExtensionAddPod(ctx, state, pod, ap, nodeInfo)
+		if !status.IsSuccess() {
+			return status.AsError()
+		}
+		return nil
+	}
+
+	podPriority := podutil.GetPodPriority(pod)
+	preemptorElasticQuotaInfo, preemptorWithElasticQuota := elasticQuotaSnapshotState.elasticQuotaInfos[pod.Namespace]
+
+	// Check if there is elastic quota in the preemptor's namespace.
+	if preemptorWithElasticQuota {
+		moreThanMinWithPreemptor = preemptorElasticQuotaInfo.overUsed(preFilterState.Resource, preemptorElasticQuotaInfo.Min, resourceAdd)
+	}
+
+	// sort the pods in node by the priority class
+	sort.Slice(nodeInfo.Pods, func(i, j int) bool { return !util.MoreImportantPod(nodeInfo.Pods[i].Pod, nodeInfo.Pods[j].Pod) })
+	for _, p := range nodeInfo.Pods {
+		pElasticQuotaInfo, pWithElasticQuota := elasticQuotaSnapshotState.elasticQuotaInfos[p.Pod.Namespace]
+		// Preemption only happens between in the same type of namespace: with or without.
+		if preemptorWithElasticQuota == pWithElasticQuota {
+			// If there is elastic quota in the preemptor's namespace.
+			if preemptorWithElasticQuota {
+				if moreThanMinWithPreemptor {
```

Comment:
The logic is able to come down here is due to one of the following reasons:

1. the pod requests too much RQ, which is larger than the PQ's `max`. In this case, the pod should try to preempt pods in the same namespace
1. the pod's request results in global RQ usage greater than the global RQ's `min`. In this case, any pod subjects to RQ may be preemptable
1. (1) and (2) may happen at the same time

So we should firstly identify the exact reason the logic reaches here. The current if...else block seems not to have covered all 3 cases above.

Reply:
case 1 is right. 
case2: Not any pod subjects to RQ. I think it is just the pods which belong to ElasticQuota which eq.used > eq.min.  The pods belong to ElasticQuota which eq.used < eq.min is guaranteed.

Reply:
> case2: Not any pod subjects to RQ. I think it is just the pods which belong to ElasticQuota which eq.used > eq.min. The pods belong to ElasticQuota which eq.used < eq.min is guaranteed.

Sounds reasonable.